website/blog/2025-09-15-make-judge/index.mdx

```diff
@@ -0,0 +1,229 @@
+
+```python
+# Test the scorer on a support interaction
+result = support_scorer(
+    inputs={"issue": "Can't reset my password"},
+    outputs={"response": "Have you tried turning it off and on again?"}
+)
+
+print(f"Rating: {result.value}")
+print(f"Reasoning: {result.rationale}")
+```
+
+Output:
+
+```
+Rating: poor
+Reasoning: The response does not address the password reset issue at all.
+It provides a generic IT troubleshooting suggestion that is irrelevant to
+the customer's specific problem...
+```
+
+### Scaling Up: Evaluating Entire Datasets
+
+Testing individual responses is useful, but the real value comes when you use your scorer to evaluate entire conversation datasets. Here's how scorers integrate seamlessly with MLflow's evaluation framework:
+
+```python
+import mlflow
+import pandas as pd
+
+# Your support conversations dataset
+test_data = pd.DataFrame([
+    {
+        "inputs": {"issue": "Billing error - charged twice"},
+        "outputs": {"response": "I'll immediately refund the duplicate charge."}
+    },
+    {
+        "inputs": {"issue": "Feature request for dark mode"},
+        "outputs": {"response": "Great suggestion! I've forwarded this to our product team."}
+    }
+])
+
+# Run evaluation with your custom scorer
+results = mlflow.genai.evaluate(
+    data=test_data,
+    scorers=[support_scorer]
+)
+
+# View results
+print(results.tables["eval_results_table"])
+```
+
+## Scorers as Agents: Trace-Based Evaluation
+
+So far our scorer has evaluated simple question-answer pairs. But MLflow Scorers can act agentically—they can analyze complex AI agents that make multiple LLM calls, use tools, and follow reasoning chains.
+
+With trace-based evaluation, your scorer evaluates not just the final output, but the entire execution path:
+
+```python
+# Create a scorer that analyzes agent behavior
+agent_scorer = make_judge(
+    name="agent_efficiency",
+    instructions=(
+        "Analyze the {{ trace }} for efficiency and correctness.\n\n"
+        "Check for: unnecessary tool calls, circular reasoning, "
+        "proper error handling, and optimal execution paths.\n"
+        "Rate as: 'optimal', 'acceptable', or 'inefficient'"
+    ),
+    model="anthropic:/claude-3-opus-20240229"
+)
+
+# Evaluate a traced agent execution
+with mlflow.start_span("agent_task") as span:
+    # Your agent logic here
+    result = my_agent.execute(task)
+    trace_id = span.trace_id
+
+trace = mlflow.get_trace(trace_id)
+evaluation = agent_scorer(trace=trace)
+```
+
+## The Game Changer: Human Feedback Alignment
+
+Here's where things get interesting. What happens when your scorer disagrees with your human experts? Instead of rewriting prompts endlessly, MLflow Scorers can learn from human corrections.
+
+The process works like this:
+
+```python
+from mlflow.entities import AssessmentSource, AssessmentSourceType
+
+# Step 1: Run scorer on production data
+for trace_id in production_traces:
```

Comment:
It might not be clear how to populate `production_traces`. Let's make it the results of a `search_traces` call:
```
production_traces = search_traces(...) # Gather traces from production or development environments
```
I also don't think that we need to compute the `scorer_result`. What purpose does this serve?